copyerror.go

```diff
@@ -0,0 +1,61 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package oras
+
+import "fmt"
+
+// CopyErrorOrigin defines the source of a copy error.
+type CopyErrorOrigin string
```

Comment:
Should it just be an int?

Reply:
or simply remove this type?

Reply:
I still prefer the current implementation, because:

1. A dedicated type ensures type safety and limits the possible values
2. The `string` type is more convenient for formatting the error message (see the `Error()` function)

Reply:
Since this is library code, I would prefer an `int`-based enum with a `String()` method. It would be less error prone than `string`-based as you can have something like
```go
switch e {
case CopyErrorOriginUnknown:
  return "unknown"
case CopyErrorOriginSource:
  return "source"
case CopyErrorOriginDestination:
  return "destination"
default:
  return "invalid"
} 
```
Also, the perf is better than `string`-based error.

Reply:
Caller won't need to know the real type of an error origin, e.g.
```golang
if err = oras.CopyGraph(ctx, src, dst, desc, extendedCopyOptions.CopyGraphOptions); err != nil {
  var re *oras.CopyError
  if !errors.As(err, &re){
    switch re.Code {
      case oras.CopyErrorOriginSource:
        return ocispec.Descriptor{}, fmt.Errorf("failed to copy from %s: %w", opts.From.Reference, err)
      case oras.CopyErrorOriginDestination:
        return ocispec.Descriptor{}, fmt.Errorf("failed to copy to %s: %w", opts.To.Reference, err)
    }
  }
}

But int performs better.

Reply:
Refactored